src/main/java/meteordevelopment/meteorclient/renderer/text/Font.java

```diff
@@ -72,57 +63,120 @@ public Font(ByteBuffer buffer, int height) {
             this.ascent = ascent.get(0);
         }
 
-        for (int i = 0; i < cdata.length; i++) {
-            STBTTPackedchar.Buffer cbuf = cdata[i];
-            int offset = packRange.get(i).first_unicode_codepoint_in_range();
-
-            for (int j = 0; j < cbuf.capacity(); j++) {
-                STBTTPackedchar packedChar = cbuf.get(j);
-
-                float ipw = 1f / size; // pixel width and height
-                float iph = 1f / size;
-
-                charMap.put(j + offset, new CharData(
-                    packedChar.xoff(),
-                    packedChar.yoff(),
-                    packedChar.xoff2(),
-                    packedChar.yoff2(),
-                    packedChar.x0() * ipw,
-                    packedChar.y0() * iph,
-                    packedChar.x1() * ipw,
-                    packedChar.y1() * iph,
-                    packedChar.xadvance()
-                ));
-            }
+        // Preload basic ASCII characters
+        preloadAsciiCharacters();
+    }
+
+    private void preloadAsciiCharacters() {
+        STBTTPackedchar.Buffer cdata = STBTTPackedchar.create(128); // Basic Latin
+
+        // create the pack range
+        STBTTPackRange.Buffer packRange = STBTTPackRange.create(1);
+        packRange.put(STBTTPackRange.create().set(height, 32, null, 128, cdata, (byte) 2, (byte) 2));
+        packRange.flip();
+
+        // pack font ranges
+        STBTruetype.stbtt_PackFontRanges(packContext, buffer, 0, packRange);
+
+        // Load character data into charMap
+        for (int i = 0; i < cdata.capacity(); i++) {
+            STBTTPackedchar packedChar = cdata.get(i);
+            putCharData(i + 32, packedChar);
         }
+
+        // Create texture
+        createTexture();
+    }
+
+    private void loadCharacter(List<Integer> codePoints) {
+        if (System.currentTimeMillis()-loadTimer>100){
+            loadTimer = System.currentTimeMillis();
+            loadCount =0;
+        }
+        if (loadCount>=loadSpeedLimit)return;
+        //Limit the load speed to avoid blocking the rendering thread
+        for (Integer codePoint : codePoints) {
+            loadCharacter(codePoint);
+        }
+        // Re-create texture
+        createTexture();
+        loadCount++;
+    }
+
+    private void loadCharacter(int codePoint) {
+        if (packedChars.containsKey(codePoint)) return;
```

Comment:
the `packedChars` map is redundant, you can replace this check here with `charMap.containsKey(codePoint)`

Reply:
Is it possible to merge this project ? It's been 4 days without news